### PR TITLE
[TGBOT-280] Create Engineer endpoint

### PR DESCRIPTION
**Pull request by schneidervs**
https://schneidervs.atlassian.net/browse/TGBOT-280

### CODE REVIEW

**Comment by schneidervs** [comment]
<html>
<body>
<h1 class="build-caption page-headline" style="box-sizing: inherit; display: flex; font-weight: 700; line-height: var(--line-height-heading); margin-bottom: var(--section-padding); margin-top: 0px; font-size: 1.6rem; align-items: center; color: rgb(20, 20, 31); font-family: system-ui, &quot;Segoe UI&quot;, roboto, &quot;Noto Sans&quot;, oxygen, ubuntu, cantarell, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span class="jenkins-icon-adjacent" style="box-sizing: inherit; margin-left: 0.5rem; width: 881.141px; line-height: 2.4rem; max-width: 1200px; overflow: hidden; text-overflow: ellipsis;">Сборка #8 (18 февр. 2023 г., 15:22:38)</span></h1><div style="box-sizing: inherit; color: rgb(20, 20, 31); font-family: system-ui, &quot;Segoe UI&quot;, roboto, &quot;Noto Sans&quot;, oxygen, ubuntu, cantarell, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div id="description" style="box-sizing: inherit;"><div style="box-sizing: inherit;"></div><div class="jenkins-buttons-row jenkins-buttons-row--invert" style="box-sizing: inherit; align-items: center; display: flex; justify-content: flex-end;"><a class="jenkins-button jenkins-button--tertiary" id="description-link" href="http://3.124.0.23:8080/job/rest-api/8/editDescription" onclick="return replaceDescription();" style="box-sizing: inherit; font-weight: 500; text-decoration: none !important; --item-background--hover:var(--button-background--hover); --item-background--active:var(--button-background--active); --item-box-shadow--focus:var(--button-box-shadow--focus); align-items: center; appearance: none; background: transparent; border: none; border-radius: 0.66rem; cursor: pointer; display: inline-flex; font-size: 0.8rem; gap: 1ch; justify-content: center; margin-top: 0px; margin-right: 0px !important; margin-bottom: 0px; margin-left: 0.5rem !important; min-height: 36px; outline: none; padding: 0.5rem 0.85rem; position: relative; white-space: nowrap; z-index: 0; color: var(--link-color);"><svg class="" aria-hidden="true" width="512px" height="512px" viewBox="0 0 512 512" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><path d="M399.608914,57 C413.784791,57 427.960684,62.4078333 438.776426,73.2235621 C449.592141,84.0392638 455,98.2149574 455,112.390686 C455,126.566468 449.592105,140.742274 438.776426,151.558078 L438.776426,151.558078 L191.040603,399.293596 C182.232434,408.101755 171.575528,414.840664 159.841736,419.022244 L159.841736,419.022244 L58.9309718,454.983885 C58.3980325,455.005817 57.9083807,454.793595 57.5574476,454.442654 C57.2063668,454.091565 56.9941379,453.601684 57.0161199,453.068522 L57.0161199,453.068522 L92.977296,352.157786 C97.1588787,340.423838 103.897856,329.766792 112.706129,320.958529 L112.706129,320.958529 L360.441401,73.2235621 C371.257143,62.4078333 385.433036,57 399.608914,57 Z" stroke="currentColor" stroke-width="32" fill-rule="nonzero"></path><polyline fill="currentColor" transform="translate(362.692388, 154.192388) rotate(45.000000) translate(-362.692388, -154.192388) " points="308.192388 138.192388 359.945436 138.192388 417.192388 138.192388 417.192388 170.192388 360.652543 170.192388 308.192388 170.192388"></polyline></g></svg>добавить описание</a></div></div></div>

  | No changes.
-- | --
  | Создана пользователем Vladimir_Schneider
  | Revision: f6f574297baef26ce7df78429dfbf40875955d3cRepository: https://github.com/Tutritp01/API-servise-TgBot.gitrefs/remotes/origin/TGBOT-280-create-engineer-endpoint
  | Результаты теста (Нет ошибок.)


</body>
</html>Сборка #8 (18 февр. 2023 г., 15:22:38)
[добавить описание](http://3.124.0.23:8080/job/rest-api/8/editDescription)
No changes.
Создана пользователем [Vladimir_Schneider](http://3.124.0.23:8080/user/vladimir_schneider)

	Revision: f6f574297baef26ce7df78429dfbf40875955d3c
Repository: https://github.com/Tutritp01/API-servise-TgBot.git
refs/remotes/origin/TGBOT-280-create-engineer-endpoint
[Результаты теста](http://3.124.0.23:8080/job/rest-api/8/testReport/) (Нет ошибок.)
Сборка модулей
 [sto-rest-api-service](http://3.124.0.23:8080/job/rest-api/8/com.tutrit$API-service-TgBot/)	0.81 секунд
 [persistence.core](http://3.124.0.23:8080/job/rest-api/8/com.tutrit$persistence-core/)	6.3 секунд
 [rest-api-service](http://3.124.0.23:8080/job/rest-api/8/com.tutrit$rest-api-service/)	8 секунд
 [util](http://3.124.0.23:8080/job/rest-api/8/com.tutrit$util/)	0.77 секунд

**Review by Artem-Metlushko — APPROVED** [review]

File: in-memory-repository/src/main/java/com/tutrit/persistence/MemoryEngineerPersistence.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import com.tutrit.persistence.core.bean.Engineer;
+import com.tutrit.persistence.core.persistence.EngineerPersistence;
+import com.tutrit.persistence.core.service.UUIDWrapper;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class MemoryEngineerPersistence implements EngineerPersistence {
+    private final UUIDWrapper uuid;
```
Comment (Sav4ykVladzislau) [inline in-memory-repository/src/main/java/com/tutrit/persistence/MemoryEngineerPersistence.java:12]: The implementation of this function is not provided by the task
https://schneidervs.atlassian.net/browse/TGBOT-280

Comment (schneidervs) [reply in-memory-repository/src/main/java/com/tutrit/persistence/MemoryEngineerPersistence.java:12]: It's a simple wrapper.

**Review by yurasava — DISMISSED** [review]

**Review by AlexanderYarokhovich — APPROVED** [review]

File: rest-api-service/src/test/java/com/tutrit/storestapiservice/client/EngineerClientTest.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+@SpringBootTest
+class EngineerClientTest {
+    @Autowired
+    EngineerClient engineerClient;
+    @MockBean
+    EngineerPersistence engineerPersistence;
+
+    @Test
+    void save() {
```
Comment (max-java) [inline rest-api-service/src/test/java/com/tutrit/storestapiservice/client/EngineerClientTest.java:21]:
```
    @Test
    void whenNewEngineerSaveShouldSave() {

    @Test
    void whenUpdatingEngineer_SaveShouldUpdate() {
```

File: rest-api-service/src/test/java/com/tutrit/storestapiservice/client/EngineerClientTest.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    @Test
+    void findById() {
+        when(engineerPersistence.findById("2")).thenReturn(makeExpected());
+        var actualEngineer = engineerClient.findById("2");
+        assertEquals(makeExpected(), actualEngineer);
+    }
+
+    private Engineer makeDebut() {
+        var engineer = new Engineer();
+        engineer.setId("1");
```
Comment (max-java) [inline rest-api-service/src/test/java/com/tutrit/storestapiservice/client/EngineerClientTest.java:36]: You can use just makeEngineer(String id) method to create a victim
```java
private Engineer makeEngineer(String id) {
  var eng = new Engineer();
  eng.setId(id);
  return eng;
```

File: rest-api-service/src/test/java/com/tutrit/storestapiservice/service/EngineerServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Test
+    void findById() {
+        when(engineerClient.findById("2")).thenReturn(makeExpected());
+        var actualEngineer = engineerService.findById("2");
+        assertEquals(makeExpected(), actualEngineer);
+    }
+
+    @Test
+    void findByIdNull() {
```
Comment (max-java) [inline rest-api-service/src/test/java/com/tutrit/storestapiservice/service/EngineerServiceTest.java:37]: Better name would be whenEngineerNotFoundShouldReturnNull (),
But we want NotFoundException.

**Review by max-java — CHANGES_REQUESTED** [review]
ptal

**Review by Sav4ykVladzislau — APPROVED** [review]
lgfm

**Review by max-java — CHANGES_REQUESTED** [review]
ptal

**Review by yurasava — APPROVED** [review]

**Comment by max-java** [comment]
any updates on this?

**Review by AlexanderYarokhovich — APPROVED** [review]

**Review by Artem-Metlushko — APPROVED** [review]

**Review by Sav4ykVladzislau — APPROVED** [review]

**Comment by max-java** [comment]
any updates on this?

**Pull request closed without merging**